services/api/src/models/__tests__/audit-log.js

```diff
@@ -0,0 +1,98 @@
+const AuditLog = require('../audit-log');
+const User = require('../user');
+const Koa = require('koa');
+const Router = require('@koa/router');
+const request = require('supertest');
+
+const { setupDb, teardownDb } = require('../../utils/testing');
+
+beforeAll(async () => {
+  await setupDb();
+});
+
+afterAll(async () => {
+  await teardownDb();
+});
+
+async function getContext(user) {
+  const app = new Koa();
+  const productRouter = new Router({
+    prefix: '/1/products',
+  });
+  let current;
+
+  productRouter.use((ctx, next) => {
+    ctx.state.authUser = user;
+    return next();
+  });
+  productRouter.get('/:id', async (ctx) => {
+    current = ctx;
+  });
+  app.use(productRouter.routes());
+  await request(app.callback()).get('/1/products/id');
+
+  return current;
+}
+
+describe('AuditLog', () => {
+  describe('getDiffObject', () => {
+    it('should return a diff object', async () => {
+      const user = new User({});
+      user.email = 'bob@new.com';
+      const diffObject = AuditLog.getDiffObject(user, ['email']);
+
+      expect(JSON.stringify(diffObject)).toBe(
+        JSON.stringify({
+          email: 'bob@new.com',
+        })
+      );
+    });
+  });
+
+  describe('getFieldsContext', () => {
+    it('should extract fields from ctx', async () => {
+      const user = new User({});
+      const ctx = await getContext(user);
+
+      expect(AuditLog.getFieldsContext(ctx)).toEqual(
+        expect.objectContaining({
+          requestMethod: 'GET',
+          requestUrl: '/1/products/id',
+          routeNormalizedPath: '/1/products/:id',
+          routePrefix: '/1/products',
+          user: user.id,
+        })
+      );
+    });
+  });
+
+  describe('append', () => {
+    it('should write to the db', async () => {
+      const user = new User({ email: 'bob@gmail.com' });
+      const ctx = await getContext(user);
+
+      await AuditLog.append('did something', {
+        type: 'security',
+        ctx,
+        objectDiff: { email: user.email },
```

Comment:
I'm slightly confused by the objectDiff as you have it now.

What we want is:

- For certain API calls, we want to log if certain attributes were changed and what those attributes were changed to
- So for example, if we have a `ClinicalStudy` object with a `numberOfDays` attribute. We would want to specify that we want to track `numberOfDays` and we would want to know the BEFORE and the AFTER state of that attribute.
- This would show up in the audit log as: "Sarah changed Clinical Study X: numberOfDays was changed from 500 to 600"

I'm not sure if we're capturing this here

Reply:
I now have `objectBefore` and `objectAfter` to capture that information.